WalletWasabi.Fluent/Controls/TagsBox.xaml.cs

```diff
@@ -0,0 +1,325 @@
+            set => SetValue(RestrictInputToSuggestionsProperty, value);
+        }
+
+        public object SelectedTag
+        {
+            get => GetValue(SelectedTagProperty);
+            set => SetValue(SelectedTagProperty, value);
+        }
+
+        public int ItemCountLimit
+        {
+            get => GetValue(ItemCountLimitProperty);
+            set => SetValue(ItemCountLimitProperty, value);
+        }
+
+        public IEnumerable? Suggestions
+        {
+            get => _suggestions;
+            set => SetAndRaise(SuggestionsProperty, ref _suggestions, value);
+        }
+
+        private void CheckIsInputEnabled()
+        {
+            if (Items is { } && ItemCountLimit > 0 && Items is IList x)
+            {
+                _isInputEnabled = x.Count < ItemCountLimit;
+            }
+        }
+
+        protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e)
+        {
+            base.OnDetachedFromVisualTree(e);
+            _disposable?.Dispose();
+        }
+
+        protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+        {
+            base.OnApplyTemplate(e);
+
+            Presenter.ApplyTemplate();
+
+            _autoCompleteBox = (Presenter.Panel as ConcatenatingWrapPanel)?.ConcatenatedChildren
+                .OfType<AutoCompleteBox>().FirstOrDefault();
+
+            if (_autoCompleteBox is null) return;
+
+            _autoCompleteBox.KeyUp += OnKeyUp;
+            _autoCompleteBox.TextChanged += OnTextChanged;
+            _autoCompleteBox.DropDownClosed += OnDropDownClosed;
+
+            _autoCompleteBox.GotFocus += OnOnAutoCompleteBoxGotFocus;
+            _autoCompleteBox.LostFocus += OnAutoCompleteBoxLostFocus;
+
+            _disposable =
+                _autoCompleteBox.AddDisposableHandler(TextInputEvent, OnTextInput,
+                    RoutingStrategies.Tunnel);
+        }
+
+        private void OnAutoCompleteBoxLostFocus(object? sender, RoutedEventArgs e)
+        {
+            PseudoClasses.Set(":focus", false);
+        }
+
+        private void OnOnAutoCompleteBoxGotFocus(object? sender, GotFocusEventArgs e)
+        {
+            PseudoClasses.Set(":focus", true);
+        }
+
+        private void OnTextInput(object? sender, TextInputEventArgs e)
+        {
+            if (_autoCompleteBox is null)
+            {
+                return;
+            }
+
+            if (!_isInputEnabled)
+            {
+                e.Handled = true;
+                return;
+            }
+
+            if (RestrictInputToSuggestions &&
+                Suggestions is { } &&
+                !Suggestions.Cast<string>().Any(x => x.StartsWith(_autoCompleteBox.SearchText ?? "", true, CultureInfo.CurrentCulture)))
+            {
+                e.Handled = true;
+            }
+        }
+
+        protected override void UpdateDataValidation<T>(AvaloniaProperty<T> property, BindingValue<T> value)
```

Comment:
What does this do?

Reply:
this is needed for data validations system to work, passing down error notifications to an avaloniaprop.. (and also copied it from ACB)

Reply:
ok